runtime/parachains/src/ump.rs

```diff
@@ -45,45 +44,56 @@ pub trait UmpSink {
 	/// it did not begin processing a message since it would otherwise exceed `max_weight`.
 	///
 	/// See the trait docs for more details.
-	fn process_upward_message(origin: ParaId, msg: &[u8], max_weight: Weight) -> Option<Weight>;
+	fn process_upward_message(origin: ParaId, msg: &[u8], max_weight: Weight) -> Result<Weight, (MessageId, Weight)>;
 }
 
 /// An implementation of a sink that just swallows the message without consuming any weight. Returns
 /// `Some(0)` indicating that no messages existed for it to process.
 impl UmpSink for () {
-	fn process_upward_message(_: ParaId, _: &[u8], _: Weight) -> Option<Weight> {
-		Some(0)
+	fn process_upward_message(_: ParaId, _: &[u8], _: Weight) -> Result<Weight, (MessageId, Weight)> {
+		Ok(0)
 	}
 }
 
+/// Simple type used to identify messages for the purpose of reporting events. Secure if and only
+/// if the message content is unique.
+pub type MessageId = [u8; 32];
+
 /// A specific implementation of a UmpSink where messages are in the XCM format
 /// and will be forwarded to the XCM Executor.
-pub struct XcmSink<XcmExecutor, Call>(PhantomData<(XcmExecutor, Call)>);
+pub struct XcmSink<XcmExecutor, Config>(PhantomData<(XcmExecutor, Config)>);
 
-impl<XcmExecutor: xcm::v0::ExecuteXcm<Call>, Call> UmpSink for XcmSink<XcmExecutor, Call> {
-	fn process_upward_message(origin: ParaId, mut msg: &[u8], max_weight: Weight) -> Option<Weight> {
+impl<XcmExecutor: xcm::v0::ExecuteXcm<C::Call>, C: Config> UmpSink for XcmSink<XcmExecutor, C> {
+	fn process_upward_message(origin: ParaId, data: &[u8], max_weight: Weight) -> Result<Weight, (MessageId, Weight)> {
 		use parity_scale_codec::Decode;
 		use xcm::VersionedXcm;
-		use xcm::v0::{Junction, MultiLocation, Outcome, Error as XcmError};
-
-		if let Ok(versioned_xcm_message) = VersionedXcm::decode(&mut msg) {
-			match versioned_xcm_message {
-				VersionedXcm::V0(xcm_message) => {
-					let xcm_junction: Junction = Junction::Parachain(origin.into());
-					let xcm_location: MultiLocation = xcm_junction.into();
-					match XcmExecutor::execute_xcm(xcm_location, xcm_message, max_weight) {
-						Outcome::Complete(w) | Outcome::Incomplete(w, _) => Some(w),
-						Outcome::Error(XcmError::WeightLimitReached(..)) => None,
-						Outcome::Error(_) => Some(0),
+		use xcm::v0::{Xcm, Junction, MultiLocation, Error as XcmError};
+
+		let id = sp_io::hashing::blake2_256(&data[..]);
```

Comment:
Because `decode` allows for unconsumed trailing data, this may lead to the situation where there are two values `x` and `y` of type `VersionedXcm` and `x == y` but `id(x) ≠ id(y)`. Not sure if that's a problem, but `MessageId` docs may imply that it is.

Reply:
id is 256 bit which will always fit into [u8; 32], so not sure what you mean here actually

Reply:
Both `u8::decode(&mut &[69])` and `u8::decode(&mut &[69, 0])` return `69` as a result. `blake2_256(&[69])` and `blake2_256(&[69, 0])` obviously return different results.

Reply:
Not really - the docs mean that the same `MessageId` must imply the same message. They don't imply the reverse.